### DISCUSSION TITLE
azd up support for App Service Environment

### DISCUSSION THREAD

**Discussion by tylerj-aft** [Azure/azure-dev#889] [Ideas]
Will azd up support deployment into an App Service Environment? If yes, when?

**Comment by rajeshkamal5050** [comment]
@tylerj-aft Azd has been supporting **_App Service hosts_** for some time now - https://learn.microsoft.com/en-us/azure/developer/azure-developer-cli/overview#supported-azure-compute-services-host

<img width="598" alt="Azd_Supported_Hosts_Langs" src="https://user-images.githubusercontent.com/11532743/195467679-e1334e78-88bf-4b15-bcf1-a9243750dd8d.png">

Just curious, on whether you had a chance to take a look at our docs - https://learn.microsoft.com/en-us/azure/developer/azure-developer-cli/overview

Please give it a try and let us know.

**Reply by tylerj-aft** [reply]
@rajeshkamal5050 I'm aware that it supports app service hosts. But after testing it today, it does not support deploying to an app service environment. See https://learn.microsoft.com/en-us/azure/app-service/environment/overview to understand what an ASE is. It should be available in the region list when one has been created in your subscription. It's not listed in the region list. It also does not allow you to name your ASE in the .env file without throwing an exception when calling azd up.

**Reply by rajeshkamal5050** [reply]
Ah I see. Thanks for the pointer. Can you file a feature request here https://github.com/Azure/azure-dev/issues for the same?

**Reply by Glober777** [reply]
Actually, it may only seem like the `location` parameter should get the name of the ASE (because that's how it looks like in Azure Portal). However, when deploying an App Service (i.e. Web App) into ASE, the `location` parameter is actually populated with a value from the list of regions (i.e. eastus, westeu, etc.), but there is also a `hostingEnvironment` property (under `properties`) which is be populated with the resource ID of the ASE. (Ref: [Microsoft.Web sites](https://learn.microsoft.com/en-us/azure/templates/microsoft.web/sites?pivots=deployment-language-bicep#resource-format-1)). You can check that by going through the new Web Site wizard in Azure Portal and pressing `Download a template for automation` link on the last page (it will show you the actual ARM template for what you're about to provision).

With that, it is actually not a problem to update the bicep code to provision the app service resource under ASE. The actual problem is with deploying the compiled app code (i.e. ZIP deploy) to a provisioned App Service resource (hosted on ASE), since App Service instances in ASE get a different endpoint address (i.e. <appName>.scm.<ASE-name>.ammpserviceenvironment.net and <appName>.<ASE-name>.ammpserviceenvironment.net). And this is not only related to Web Sites, but to other services that can be deployed into ASE-hosted App Service (i.e. Function Apps, Logic Apps, Docker Containers).

To me it seems like this could be solved with an additional parameter that would be taking the custom endpoint name as a value for deployment. Potentially, this could be something that will get populated with an output from Provisioning task.
